url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/CpuCore.java

```diff
@@ -0,0 +1,2574 @@
+		reg_pc += 2;
+		return value;
+	}
+
+	public int fetchPCWordSigned()
+	{
+		int value = readMemoryWordSigned(reg_pc);
+		reg_pc += 2;
+		return value;
+	}
+
+	public int fetchPCLong()
+	{
+		int value = readMemoryLong(reg_pc);
+		reg_pc += 4;
+		return value;
+	}
+
+	// status reg
+	public int getCCRegister()
+	{
+		return reg_sr & 0x00ff;
+	}
+
+	public int getSR()
+	{
+		return reg_sr;
+	}
+
+	public void setCCRegister(int value)
+	{
+		reg_sr = (reg_sr & 0xff00) | (value & 0x00ff);
+	}
+
+	public void setSR(int value)
+	{
+		//check for supervisor bit change
+		if(((reg_sr & SUPERVISOR_FLAG) ^ (value & SUPERVISOR_FLAG)) != 0)
+		{
+			//if changing via this method don't push/pop sr and pc - this is only called by andi/eori/ori
+
+			if((value & SUPERVISOR_FLAG) != 0)
+			{
+				reg_usp = addr_regs[7];
+				addr_regs[7] = reg_ssp;
+			}
+			else
+			{
+				//switch stacks
+				reg_ssp = addr_regs[7];
+				addr_regs[7] = reg_usp;
+			}
+		}
+		reg_sr = value;
+	}
+
+	/**
+	 * Set the SR when coming from an RTE.
+	 * We might already have been IN supervisor mode when the exception was caused (eg a Trap called in supervisor mode),
+	 * so we must check the S bit we get back from the stack and possibly STAY in supervisor mode even after the RTE.
+	 * @param value
+	 */
+	public void setSR2(int value)
+	{
+		// old value of SR, this will be in supermode
+		reg_sr = value;							// new value of SR, could be user mode or super mode
+		if ((reg_sr & SUPERVISOR_FLAG) == 0)	// we changed back to user mode,change stack pointer
+		{
+			reg_ssp = addr_regs[7];				// keep supervisor stack pointer
+			addr_regs[7] = reg_usp ;			// get user stack pointer
+		}
+	}
+
+	public void setFlags(int flags)
+	{
+		//only set CC flags so clear top byte
+		reg_sr |= (flags & 0x00ff);
+	}
+
+	public void clrFlags(int flags)
+	{
+		//we need to invert the flags and then AND them with the SR to clear
+		reg_sr &= ~(flags & 0x00ff);
+	}
+
+	public boolean isFlagSet(int flag)
+	{
+		return ((reg_sr & flag) == flag);
+	}
+
+	public void calcFlags(InstructionType type, int src, int dst, int result, Size sz)
+	{
+		calcFlagsParam(type, src, dst, result, 0, sz);
+	}
+
+	public void calcFlagsParam(InstructionType type, int src, int dst, int result, int extraParam, Size sz)
+	{
+		boolean Sm = (src & sz.msb()) != 0;
+		boolean Dm = (dst & sz.msb()) != 0;
+		boolean Rm = (result & sz.msb()) != 0;
+
+		switch(type)
+		{
+			case ADD:	//ADD, ADDI, ADDQ
+			{
+				if((Sm && Dm && !Rm) || (!Sm && !Dm && Rm))
+				{
+					reg_sr |= V_FLAG;
+				}
+				else
+				{
+					reg_sr &= ~(V_FLAG);
+				}
+
+				if((Sm && Dm) || (!Rm && Dm) || (Sm && !Rm))
+				{
+					reg_sr |= (C_FLAG | X_FLAG);
+				}
+				else
+				{
+					reg_sr &= ~(C_FLAG | X_FLAG);
+				}
+
+				if(result == 0)
+				{
+					reg_sr |= Z_FLAG;
+				}
+				else
+				{
+					reg_sr &= ~(Z_FLAG);
+				}
+
+				if(Rm)
+				{
+					reg_sr |= N_FLAG;
+				}
+				else
+				{
+					reg_sr &= ~(N_FLAG);
+				}
+				break;
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 299:336, 669:706.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.